lib/src/util/dart_type_utilities.dart

```diff
@@ -13,6 +13,9 @@ import 'package:analyzer/dart/element/type.dart';
 typedef bool AstNodePredicate(AstNode node);
 
 class DartTypeUtilities {
+  static bool areFromSameClass(DartType type1, DartType type2) =>
+      type2 != null && isClass(type1, type2.name, type2.element?.library?.name);
```

Comment:
I think for these use cases you really want to know whether whether the types are the same. That is, I don't think you want to suggest dropping the type if it's the same class but with a different type argument:
```
List<num> x = <int>[];
```
If that's true, then you can just use the `==` operator on the types. If that's not the case (and you should add the above as a test case), then you can just ask whether the elements associated with the types are the same.